### PR TITLE
fix: improve overall run-ios experience and get rid of CFBundleIdentifier

### LINKED ISSUE #107: Get rid of "CFBundleIdentifier" error
This is super old issue that happens when you run "run-ios" (usually for the first time on a newly inited project). It's one of the pinned issues in React Native and one of the most popular errors people are struggling with.

If this is a "wont fix", we should consider changing the logic of the command itself to avoid this mistake. 

I am opening this issue as a memo for the future.

Related: https://github.com/facebook/react-native/issues/21303

Stack trace:
```
Error: Command failed: /usr/libexec/PlistBuddy -c Print:CFBundleIdentifier build/Build/Products/Debug-iphonesimulator/RNTesterPods.app/Info.plist
Print: Entry, ":CFBundleIdentifier", Does Not Exist

    at checkExecSyncError (child_process.js:611:11)
    at Object.execFileSync (child_process.js:629:13)
    at Promise.then (/Users/grabbou/Repositories/react-native/node_modules/@react-native-community/cli/build/runIOS/runIOS.js:183:36)
```

### PR DESCRIPTION
**Pull request by grabbou**
Fixes #107 

Summary:
---------

Bugs fixed:
- "CFBundleIdentifier" printed every time "run-ios" fails, shadowing the real cause for the problem
- "run-ios" running a previous/cached version of an app (usually when there was an error during build step that was ignored)

Changes made:
- Decided to turn Flow and ESLint and apply all the changes as I am working on it. I know the diff might be strange at first sight, but if you look closely, it's just cosmetics to satisfy various rules
- I also extracted some blocks to separate utilities for better readability
- I also removed a no longer needed check for CRNA which is deprecated (and IMO, is not needed)
- Converted few functions to "async" form for better readability
- Started capturing the "stderr" of "xcodebuild" and started relying on its error code in order to avoid going to next step when it fails
- Added a new type of error to differentiate from other errors and also be able to capture additional data. Logic and styling subject to change, encapsulated inside Logger.

Test Plan:
----------

Here's how the error messages look before/now:

While try to build a project with a wrong "schema" :

Before:
screenshot 2019-01-31 at 18 07 43

After:
screenshot 2019-01-31 at 18 05 23

While building a project that has a typo in "AppDelegate.m":

Before:
screenshot 2019-01-31 at 18 08 19

After:
screenshot 2019-01-31 at 18 06 36

You can try it out by running `run-ios` command on a broken project (typo, wrong configuration) or by passing a non-existing `schema`

### CODE REVIEW

File: packages/cli/src/runIOS/runIOS.js [hunk 1]
```diff
@@ -308,28 +307,77 @@ function buildProject(
...
-        productNameMatch.length > 1
-      ) {
-        return resolve(productNameMatch[1]); // 0 is the full match, 1 is the app name
+      if (code !== 0) {
+        reject(
+          new ProcessError(
+            [
+              `Failed to build iOS project.`,
+              `We ran "xcodebuild" command but it exited with error code ${code}.`,
+              `Consider building your app with XCode app to further debug this issue.`,
```
Comment (eliperkins) [inline packages/cli/src/runIOS/runIOS.js]: s/XCode/Xcode

Comment (eliperkins) [reply packages/cli/src/runIOS/runIOS.js]: What do you think about phrasing this as: `To debug build logs furher, consider building your app with Xcode.app, by opening ${xcodeProject}`?

Comment (grabbou) [reply packages/cli/src/runIOS/runIOS.js]: Love that! I was struggling with error messages in particular as I always have a hard time phrasing it in a super friendly way! I'll definitely use that one!

**Comment by elucaswork** [comment]
Nice work! 👏  

> Started capturing the "stderr" of "xcodebuild" and started relying on its error code in order to avoid going to next step when it fails

Cool, so this will prevent us to have a `"main.jsbundle does not exist."` error, right? 🙌(I've suffered with it this week - https://stackoverflow.com/a/54464135/1394414)

File: packages/cli/src/util/logger.js [hunk 2]
```diff
@@ -23,13 +23,32 @@ const error = (...messages: Array<string>) => {
...
+  if (err instanceof ProcessError) {
+    console.error(`${chalk.grey(err.processError)}`);
+  }
+};
+
 const debug = (...messages: Array<string>) => {
   console.log(`${chalk.black.bgWhite(' DEBUG ')} ${joinMessages(messages)}`);
 };
 
+class ProcessError extends Error {
```
Comment (thymikee) [inline packages/cli/src/util/logger.js]: This doesn't belong to logger. Can you move it to some kind of `errorHelpers`?

File: packages/cli/src/util/logger.js [hunk 3]
```diff
@@ -23,13 +23,32 @@ const error = (...messages: Array<string>) => {
...
+
+  processError: string;
+}
+
 module.exports = {
   info,
   warn,
   error,
   debug,
+  logError,
```
Comment (thymikee) [inline packages/cli/src/util/logger.js]: maybe `logProcessError`? It's very scoped to `ProcessError`

File: packages/cli/src/runIOS/runIOS.js [hunk 4]
```diff
@@ -4,79 +4,68 @@
...
+type FlagsT = {
+  simulator: string,
+  configuration: string,
+  scheme: ?string,
+  projectPath: string,
+  device: ?string,
+  udid: ?string,
+  packager: boolean,
+  verbose: boolean,
+  port: number,
```
Comment (thymikee) [inline packages/cli/src/runIOS/runIOS.js]: This type seems wrong all the way. These are arguments, so they're all optional, right? Shouldn't it be typed like this?
```jsx
type FlagsT = {|
  simulator?: string,
  configuration?: string,
  scheme?: string,
  projectPath?: string,
  device?: string,
  udid?: string,
  packager?: boolean,
  verbose?: boolean,
  port?: number,
|};
```

File: packages/cli/src/runIOS/runIOS.js [hunk 5]
```diff
@@ -91,23 +80,31 @@ function runIOS(argv, config, args) {
...
-  } else {
-    return runOnSimulator(xcodeProject, args, scheme);
   }
+
+  return runOnSimulator(xcodeProject, args, scheme);
 }
 
-function runOnDeviceByUdid(args, scheme, xcodeProject, devices) {
+function runOnDeviceByUdid(
+  args: FlagsT & { udid: string },
```
Comment (thymikee) [inline packages/cli/src/runIOS/runIOS.js]: see above comment

File: packages/cli/src/runIOS/runIOS.js [hunk 6]
```diff
@@ -308,28 +307,79 @@ function buildProject(
...
+              `Failed to build iOS project.`,
+              `We ran "xcodebuild" command but it exited with error code ${code}.`,
+              `To debug build logs further, consider building your app with Xcode.app, by opening ${
+                xcodeProject.name
+              }`,
+            ].join(' '),
+            errorOutput
+          )
+        );
+        return;
```
Comment (thymikee) [inline packages/cli/src/runIOS/runIOS.js]: is this return necessary after we `reject`ed?

File: packages/cli/src/util/logger.js (hunk 2 above)
Comment (grabbou) [reply packages/cli/src/util/logger.js]: Yup, good idea. I'd suggest "errors" for now.

File: packages/cli/src/util/logger.js (hunk 3 above)
Comment (grabbou) [reply packages/cli/src/util/logger.js]: It is scoped for now, but the intent is to log all other kinds of errors too. We use it in the main cli interface where we catch the error from the command and don't know it's type yet.

File: packages/cli/src/runIOS/runIOS.js (hunk 4 above)
Comment (grabbou) [reply packages/cli/src/runIOS/runIOS.js]: It seems, but if you take into consideration default values set by the CLI (see the object at the end where options are defined), it looks good. In this context, e.g. `verbose` is always false, unless set. Same for `port`. We do the same in other commands.

We can add a comment if that's confusing.

File: packages/cli/src/runIOS/runIOS.js (hunk 6 above)
Comment (grabbou) [reply packages/cli/src/runIOS/runIOS.js]: It's not - just my personal preference of early returns over /else clause (visual appearance).

Comment (grabbou) [reply packages/cli/src/runIOS/runIOS.js]: I think I'll update it to if/else here 👌

File: packages/cli/src/util/logger.js (hunk 3 above)
Comment (grabbou) [reply packages/cli/src/util/logger.js]: See: https://github.com/react-native-community/react-native-cli/pull/139/files#diff-97059d274ee5d7bbcc5730ddee2a98daR30

File: packages/cli/src/runIOS/runIOS.js (hunk 4 above)
Comment (thymikee) [reply packages/cli/src/runIOS/runIOS.js]: I mean, if an option is present in the object, why would it be falsy? Maybe that's something coming from commander, I didn't check locally.

Comment (thymikee) [reply packages/cli/src/runIOS/runIOS.js]: Just checked, this is what commander returns, so feel free to discard my comment

**Comment by grabbou** [comment]
Check @thymikee if you are happy with the last commit and if yes, let's merge it :fire:

File: packages/cli/src/util/errors.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
+/**
+ * @flow
+ */
+const chalk = require('chalk');
+
+class ProcessError extends Error {
+  constructor(msg: string, processError: string) {
+    super(`${chalk.red(msg)}\n\n${chalk.gray(processError)}`);
```
Comment (thymikee) [inline packages/cli/src/util/errors.js]: Looks silly, but I'm cool with that :)

Comment (grabbou) [reply packages/cli/src/util/errors.js]: I would say it's a "temporary" solution when we figure out proper logging. I think it's a good starting point.

**Comment by owcc** [comment]
I follow the issue " ":CFBundleIdentifier", Does Not Exist"to here, and this issue is still exist🙁

**Pull request merged**